hoot-core/src/main/cpp/hoot/core/io/ApiDbReader.cpp

```diff
@@ -291,6 +292,64 @@ bool ApiDbReader::_isValidBounds(const Envelope& bounds)
   return true;
 }
 
+void ApiDbReader::_readWaysByNodeIds(OsmMapPtr map, const QSet<QString>& nodeIds, QSet<QString>& wayIds,
+                                     QSet<QString>& additionalNodeIds, long& nodeCount, long& wayCount)
+{
+  LOG_DEBUG("Retrieving way IDs referenced by the selected nodes...");
+  boost::shared_ptr<QSqlQuery> wayIdItr = _getDatabase()->selectWayIdsByWayNodeIds(nodeIds);
+  while (wayIdItr->next())
+  {
+    const long wayId = (*wayIdItr).value(0).toLongLong();
+    LOG_VART(ElementId(ElementType::Way, wayId));
+    wayIds.insert(QString::number(wayId));
+  }
+  LOG_VARD(wayIds.size());
+  if (wayIds.size() > 0)
+  {
+    LOG_DEBUG("Retrieving ways by way ID...");
+    boost::shared_ptr<QSqlQuery> wayItr =
+      _getDatabase()->selectElementsByElementIdList(wayIds, TableType::Way);
+    while (wayItr->next())
+    {
+      WayPtr way = _resultToWay(*wayItr, *map);
+      map->addElement(way);
+      LOG_VART(way);
+      wayCount++;
+    }
+
+    LOG_DEBUG("Retrieving way node IDs referenced by the selected ways...");
+    boost::shared_ptr<QSqlQuery> additionalWayNodeIdItr =
+      _getDatabase()->selectWayNodeIdsByWayIds(wayIds);
+    while (additionalWayNodeIdItr->next())
+    {
+      const long nodeId = (*additionalWayNodeIdItr).value(0).toLongLong();
+      LOG_VART(ElementId(ElementType::Node, nodeId));
+      additionalNodeIds.insert(QString::number(nodeId));
+    }
+
+    //subtract nodeIds from additionalWayNodeIds so no dupes get added
+    LOG_VARD(nodeIds.size());
+    LOG_VARD(additionalNodeIds.size());
+    additionalNodeIds = additionalNodeIds.subtract(nodeIds);
+    LOG_VARD(additionalNodeIds.size());
+
+    if (additionalNodeIds.size() > 0)
+    {
+      LOG_DEBUG(
+        "Retrieving nodes falling outside of the query bounds but belonging to a selected way...");
+      boost::shared_ptr<QSqlQuery> additionalWayNodeItr =
+        _getDatabase()->selectElementsByElementIdList(additionalNodeIds, TableType::Node);
+      while (additionalWayNodeItr->next())
+      {
+        NodePtr node = _resultToNode(*additionalWayNodeItr, *map);
+        map->addElement(node);
+        LOG_VART(node);
+        nodeCount++;
+      }
+    }
+  }
+}
+
```

Comment:
This function is yanked directly from the large deletion below to reduce cognitive complexity.